crates/l1tx/src/envelope/builder.rs

```diff
@@ -7,43 +7,61 @@ use bitcoin::{
     script::PushBytesBuf,
     ScriptBuf,
 };
-use strata_primitives::l1::payload::{L1Payload, L1PayloadType};
+use strata_primitives::{
+    l1::payload::{L1Payload, L1PayloadType},
+    params::Params,
+};
 use tracing::*;
 
-use crate::envelope::parser::{BATCH_DATA_TAG, ROLLUP_NAME_TAG, VERSION_TAG};
-
 // Generates a [`ScriptBuf`] that consists of `OP_IF .. OP_ENDIF` block
 pub fn build_envelope_script(
-    envelope_data: &L1Payload,
-    // TODO: get tagnames from config
-    rollup_name: &str,
+    params: &Params,
+    payloads: &[L1Payload],
     version: u8,
 ) -> anyhow::Result<ScriptBuf> {
+    let mut bytes = Vec::new();
+    for payload in payloads {
+        let script_bytes = build_payload_envelope(params, payload, version)?;
+        bytes.extend(script_bytes);
+    }
+    Ok(ScriptBuf::from_bytes(bytes))
+}
+
+fn build_payload_envelope(
+    params: &Params,
+    payload: &L1Payload,
+    version: u8,
+) -> anyhow::Result<Vec<u8>> {
+    let tag = get_payload_type_tag(payload.payload_type(), params)?;
     let mut builder = script::Builder::new()
         .push_opcode(OP_FALSE)
         .push_opcode(OP_IF)
-        .push_slice(PushBytesBuf::try_from(ROLLUP_NAME_TAG.to_vec())?)
-        .push_slice(PushBytesBuf::try_from(rollup_name.as_bytes().to_vec())?)
-        .push_slice(PushBytesBuf::try_from(VERSION_TAG.to_vec())?)
+        .push_slice(tag)
+        // Insert version
         .push_slice(PushBytesBuf::from([version]))
-        .push_slice(PushBytesBuf::try_from(BATCH_DATA_TAG.to_vec())?)
-        .push_int(envelope_data.data().len() as i64);
+        // Insert size
+        .push_slice(PushBytesBuf::from(
+            (payload.data().len() as u32).to_be_bytes(),
+        ));
 
-    trace!(batchdata_size = %envelope_data.data().len(), "Inserting batch data");
-    for chunk in envelope_data.data().chunks(520) {
+    // Insert actual data
+    trace!(batchdata_size = %payload.data().len(), "Inserting batch data");
+    for chunk in payload.data().chunks(520) {
         trace!(size=%chunk.len(), "inserting chunk");
```

Comment:
This is *way* into the internals.  We just have extensive-enough unit tests it so that we never have to worry about if it's correct or not.

Logs are most useful to have closer to the boundaries of components or for high level components doing significant operations the user might want to know about, where we might have to worry about unexpected interactions we want to go look at later.

Reply:
Agreed. Removed the trace here.